src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -166,38 +169,118 @@ impl BatchMetrics {
     }
 }
 
+/// Manages batches and metrics, including the small-batch optimization
+/// in `write_batches`.
 struct BatchBuilderAndMetadata<K, V, T, D>
 where
     K: Codec,
     V: Codec,
     T: Timestamp + Lattice + Codec64,
 {
-    builder: mz_persist_client::batch::BatchBuilder<K, V, T, D>,
+    builder: Either<mz_persist_client::batch::BatchBuilder<K, V, T, D>, Vec<(K, V, T, D)>>,
     metrics: BatchMetrics,
 }
 
 impl<K, V, T, D> BatchBuilderAndMetadata<K, V, T, D>
 where
-    K: Codec,
-    V: Codec,
+    K: Codec + Debug,
+    V: Codec + Debug,
     T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
 {
-    fn new(bb: mz_persist_client::batch::BatchBuilder<K, V, T, D>) -> Self {
+    fn new() -> Self {
         BatchBuilderAndMetadata {
-            builder: bb,
+            builder: Either::Right(Vec::new()),
             metrics: Default::default(),
         }
     }
+
+    async fn add<F>(&mut self, spill_to_batch_size: usize, f: F, k: K, v: V, t: T, d: D)
```

Comment:
How come the spill threshold is not part of initialization? Do we need to be providing on every add?

Reply:
Because it can change in the middle of running!